kategory/src/main/kotlin/kategory/data/WriterT.kt

```diff
@@ -2,7 +2,8 @@ package kategory
 
 @higherkind data class WriterT<F, W, A>(val MF: Monad<F>, val value: HK<F, Tuple2<W, A>>) : WriterTKind<F, W, A> {
 
-    companion object {
+    companion object : WriterTInstances {
```

Comment:
Hi @raulraja. This is a different approach @pakoito and me were talking about yesterday. I wanted to add a couple of new instances for `WriterT` and had some issues. In cats, instances are normally dependent using trait mixins. `WriterTInstances` inherits behaviors from `WriterTInstances0`, which inherits behaviors from `WriterTInstances1`... etc, and the companion object inherits from `WriterTInstances` to be able to acquire those behaviors.

This is just mimicking that in the same exact way.

The PR is Work in Progress and was opened to let @pakoito (and also you, ofc) to validate this new style, since to me it's more clear and requires less code written inside the class itself (`WriterT`).

Reply:
I misunderstood your intentions yesterday. The companion object should not inherit from WriterTInstances, instances are provided on demand instead.

Reply:
The separation of instances is great but the companion should not extend them because that breaks modularity since users can't change the companion's inheritance being a singleton once declared and compiled. For that reason we decided a few weeks back after discussion to not ever use the companions as the instance and instead obtain the default instances through auxiliary methods where each method will create the instance it refers to. The reason this is done in the companion sometimes in Scala is because the Scala compiler looks in the companion for implicit resolution but in Kotlin that is not the case.

Having the companion extend an interface that adds just factory methods to the scope just creates indirection but IMO does not solve anything in Kotlin and increases the number of methods and stack size it takes to get to an instance.

The reason why we were doing that in the companion before in Kategory was to autoregister the instances but that is being moved to the @implicit declarations that Eugenio is implementing.